File: coffee-system/src/main/kotlin/com/jesperancinha/coffee/system/manager/CoffeeProcessorImpl.java
package com.jesperancinha.coffee.system.manager;

import com.jesperancinha.coffee.system.api.concurrency.QueueCallable;
import com.jesperancinha.coffee.system.concurrency.CoffeeMainCallableImpl;
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.Coffees.Coffee;
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeMachine.PaymentTypes.Payment;
import org.jesperancinha.coffee.system.input.Employees.Employee;
import org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction;
import com.jesperancinha.coffee.system.queues.Queue;
import com.jesperancinha.coffee.system.queues.QueueCofeeImpl;
import lombok.Getter;
import lombok.experimental.Accessors;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.concurrent.Callable;

@Accessors(chain = true)
@Getter
@Service
public abstract class CoffeeProcessorImpl extends ProcessorAbstract {
    private static final Logger logger = LoggerFactory.getLogger(CoffeeProcessorImpl.class);

    @Autowired
    private QueueCofeeImpl queueCofee;

    @Autowired
    private MachineProcessorImpl machineProcessor;

    public void callMakeCoffee(
            Employee employee,
            String name,
            Coffee coffee,
            Payment payment,
            List<PostAction> postActions,
            QueueCallable parentCallable
    ) {
        final CoffeeMainCallableImpl coffeCallable = new CoffeeMainCallableImpl(employee, name, coffee, payment, postActions, machineProcessor);
        parentCallable.getAllCallables().add(coffeCallable);
    }

    @Override
    public Queue getExecutorServiceQueue() {
        return queueCofee;
    }

    @Override
    public String getExecutorName(Callable<Boolean> callable) {
        return ((CoffeeMainCallableImpl) callable).getName();
    }

    public void addQueueSize(int queueSize, String name) {
        queueCofee.setQueueSize(queueSize, name);
    }

    public void initExecutors() {
        queueCofee.initExecutors();
    }

    public void stopExectutors() {
        queueCofee.stopExecutors();
    }

    public abstract void waitForAllCalls(QueueCallable queueCallable);

    public abstract void runAllCalls(QueueCallable queueCallable);
}

File: coffee-system/src/main/kotlin/org/jesperancinha/coffee/system/manager/CoffeeProcessorImpl.kt
package org.jesperancinha.coffee.system.manager

import org.jesperancinha.coffee.system.api.concurrency.QueueCallable
import org.jesperancinha.coffee.system.concurrency.CoffeeMainCallableImpl
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeeMachine.Coffees.Coffee
import org.jesperancinha.coffee.system.input.CoffeeMachines.CoffeeMachine.PaymentTypes.Payment
import org.jesperancinha.coffee.system.input.Employees.Employee
import org.jesperancinha.coffee.system.input.Employees.Employee.Actions.PostAction
import org.jesperancinha.coffee.system.queues.Queue
import org.jesperancinha.coffee.system.queues.QueueCofeeImpl
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.concurrent.Callable

@Service
class CoffeeProcessorImpl(
    @Autowired
    private val queueCofee: QueueCofeeImpl,
    @Autowired
    private val machineProcessor: MachineProcessor
) : ProcessorAbstract() {

    fun callMakeCoffee(
        employee: Employee,
        name: String,
        coffee: Coffee,
        payment: Payment,
        postActions: List<PostAction>,
        parentCallable: QueueCallable?
    ) {
        val coffeeMainCallableImpl = CoffeeMainCallableImpl(employee, name, coffee, payment, postActions, machineProcessor)
        parentCallable?.allCallables?.add(coffeeMainCallableImpl)
    }

    override val executorServiceQueue: Queue = queueCofee

    override fun getExecutorName(callable: Callable<Boolean>): String = (callable as CoffeeMainCallableImpl).name

    fun addQueueSize(queueSize: Int, name: String) {
        queueCofee.setQueueSize(queueSize, name)
    }

    fun initExecutors() {
        queueCofee.initExecutors()
    }

    fun stopExecutors() {
        queueCofee.stopExecutors()
    }
}